src/slurm.rs

```diff
@@ -0,0 +1,1000 @@
+//! defined in [RFC 8416]. They support serialization and deserialization
+//! to JSON files as required by the RFC via _serde._
+//!
+//! [RFC 8416]: https://tools.ietf.org/html/rfc8416
+
+#![cfg(feature = "slurm")]
+
+use std::{borrow, fmt, io, ops};
+use std::convert::TryFrom;
+use std::str::FromStr;
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use crate::payload::rtr;
+use crate::payload::addr::{MaxLenPrefix, Prefix};
+use crate::payload::bgpsec::KeyIdentifier;
+use crate::payload::asn::AsId;
+
+
+//------------ SlurmFile -----------------------------------------------------
+
+/// The content of a SLURM file.
+#[derive(Clone, Debug, Default, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct SlurmFile {
+    #[serde(rename = "slurmVersion")]
+    version: SlurmVersion,
+
+    #[serde(rename = "validationOutputFilters")]
+    pub filters: ValidationOutputFilters,
+
+    #[serde(rename = "locallyAddedAssertions")]
+    pub assertions: LocallyAddedAssertions,
+}
+
+impl SlurmFile {
+    /// Creates a new slurm file from filters and assertions.
+    pub fn new(
+        filters: ValidationOutputFilters,
+        assertions: LocallyAddedAssertions,
+    ) -> Self {
+        SlurmFile {
+            version: Default::default(),
+            filters, assertions,
+        }
+    }
+
+    /// Parses a SLURM file from a reader.
+    pub fn from_reader(
+        reader: impl io::Read
+    ) -> Result<Self, serde_json::Error> {
+        serde_json::from_reader(reader)
+    }
+
+    /// Returns a string with the compact JSON representation.
+    #[allow(clippy::inherent_to_string)]
+    pub fn to_string(&self) -> String {
+        serde_json::to_string(self).expect("serialization failed")
+    }
+
+    /// Returns a string with the pretty-printed JSON representation.
+    pub fn to_string_pretty(&self) -> String {
+        serde_json::to_string_pretty(self).expect("serialization failed")
+    }
+
+    /// Writes the JSON representation to a writer.
+    pub fn to_writer(&self, writer: impl io::Write) -> Result<(), io::Error> {
+        serde_json::to_writer(writer, self).map_err(Into::into)
+    }
+
+    /// Writes the pretty-printed JSON representation to a writer.
+    pub fn to_writer_pretty(
+        &self, writer: impl io::Write
+    ) -> Result<(), io::Error> {
+        serde_json::to_writer_pretty(writer, self).map_err(Into::into)
+    }
+
+    /// Returns whether the given payload item should be dropped.
+    pub fn drop_payload(&self, payload: &rtr::Payload) -> bool {
+        self.filters.drop_payload(payload)
+    }
+}
+
+impl FromStr for SlurmFile {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_str(s)
+    }
+}
+
+
+//------------ SlurmVersion --------------------------------------------------
+
+/// The SLURM version of the file.
+///
+/// This is currently required to be 1, so this type is a unit struct (de-)
```

Comment:
The main reason was that if you want line numbers in your error output – and I think having them is important –, you have to produce your error in the Serde code. If the version is wrong, then I can’t really say anything about the rest of the data since I don’t understand the data. In theory, version 2 could change everything.